src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11767,62 +11767,62 @@ private bool CanStartConsequenceExpression()
+                while (this.CurrentToken.Kind == SyntaxKind.ExclamationToken)
+                    expr = _syntaxFactory.PostfixUnaryExpression(SyntaxKind.SuppressNullableWarningExpression, expr, EatToken());
 
                 switch (this.CurrentToken.Kind)
                 {
+                    // a?.b(...)
                     case SyntaxKind.OpenParenToken:
-                        expr = _syntaxFactory.InvocationExpression(expr, this.ParseParenthesizedArgumentList());
-                        continue;
+                        return _syntaxFactory.InvocationExpression(expr, this.ParseParenthesizedArgumentList());
 
+                    // a?.b[...]
                     case SyntaxKind.OpenBracketToken:
-                        expr = _syntaxFactory.ElementAccessExpression(expr, this.ParseBracketedArgumentList());
-                        continue;
+                        return _syntaxFactory.ElementAccessExpression(expr, this.ParseBracketedArgumentList());
 
+                    // a?.b.
                     case SyntaxKind.DotToken:
-                        expr = _syntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expr, this.EatToken(), this.ParseSimpleName(NameOptions.InExpression));
-                        continue;
+                        return _syntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expr, this.EatToken(), this.ParseSimpleName(NameOptions.InExpression));
 
-                    case SyntaxKind.QuestionToken:
-                        return !CanStartConsequenceExpression()
-                            ? expr
-                            : _syntaxFactory.ConditionalAccessExpression(
-                                expr,
-                                operatorToken: this.EatToken(),
-                                ParseConsequenceSyntax());
-
-                    default:
-                        return expr;
+                    // a?.b?
+                    case SyntaxKind.QuestionToken when CanStartConsequenceExpression():
+                        return _syntaxFactory.ConditionalAccessExpression(
+                            expr,
+                            operatorToken: this.EatToken(),
+                            ParseConsequenceSyntax());
                 }
-            }
 
-            bool isOptionalExclamationsFollowedByConditionalOperation()
-            {
-                var index = 0;
-                while (this.PeekToken(index).Kind == SyntaxKind.ExclamationToken)
-                    index++;
+                // a?.b = c
+                var (operatorTokenKind, operatorExpressionKind) = GetExpressionOperatorTokenKindAndExpressionKind();
+                if (IsExpectedAssignmentOperator(operatorTokenKind))
+                {
+                    return ParseAssignmentOperator(operatorExpressionKind, expr, EatExpressionOperatorToken(operatorTokenKind));
```

Comment:
ParseAssignmentOperator is a weird name.  I would expect ParseAssignmentExpression